### PR TITLE
Initial introduction of Blink Launch Process data into Chromestatus.

### PR DESCRIPTION
**Pull request by jstenback**
Add a set of fields to the entry creation/edit form so that we collect
all the information that is part of the blink launch process. This
also makes chromestatus generate up to date intent templates with the
future intent of eliminating the need to manually create and send
intent emails. This includes making feature entries aware of the
current intent stage they're in as well as holding on to all the
additional data so that creating intent messages for later stages of
the launch process is a matter of updating the entry and generating a
new intent message.

### CODE REVIEW

File: admin.py [hunk 1]
```diff
@@ -407,7 +440,12 @@ def post(self, path, feature_id=None):
       feature = models.Feature(
           category=int(self.request.get('category')),
           name=self.request.get('name'),
+          intent_stage=int(self.request.get('intent_stage')),
           summary=self.request.get('summary'),
+          intent_to_implement_url=intent_to_implement_url,
+          origin_trial_feedback_url=origin_trial_feedback_url,
+          motivation=self.request.get('motivation'),
+          explainer_links = explainer_links,
```
Comment (ebidel) [inline admin.py]:
```suggestion
          explainer_links=explainer_links,
```

File: admin.py [hunk 2]
```diff
@@ -418,22 +456,42 @@ def post(self, path, feature_id=None):
...
-          web_dev_views=int(self.request.get('web_dev_views')),
           doc_links=doc_links,
           sample_links=sample_links,
           search_tags=search_tags,
+          comments=self.request.get('comments'),
+          experiment_goal=self.request.get('experiment_goal'),
+          experiment_timeline=self.request.get('experiment_timeline'),
+          experiment_risks=self.request.get('experiment_risks'),
+          experiment_extension_reason=self.request.get('experiment_extension_reason'),
+          ongoing_constraints = self.request.get('ongoing_constraints')
```
Comment (ebidel) [inline admin.py]:
```suggestion
          ongoing_constraints = self.request.get('ongoing_constraints'),
```

File: models.py [hunk 3]
```diff
@@ -65,6 +65,23 @@
...
+INTENT_SHIP = 5
+INTENT_REMOVE = 6
+
+INTENT_STAGES = {
+  INTENT_IMPLEMENT: 'Implement',
+  INTENT_EXPERIMENT: 'Experiment',
+  INTENT_EXTEND_TRIAL: 'Extend Origin Trial',
+  INTENT_IMPLEMENT_SHIP: 'Implement and Ship',
+  INTENT_SHIP: 'Ship',
+  INTENT_REMOVE: 'Remove'
```
Comment (ebidel) [inline models.py]:
```suggestion
  INTENT_REMOVE: 'Remove',
```

File: models.py [hunk 4]
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
...
+  owner = forms.CharField(initial=users.get_current_user().email, required=True, label='Contact emails',
       help_text='Comma separated list of full email addresses. Prefer @chromium.org.')
 
+  summary = forms.CharField(label='Feature summary', required=True, max_length=500,
+      widget=forms.Textarea(attrs={'cols': 50, 'placeholder': 'Summary description', 'maxlength': 500}))
+
+  motivation = forms.CharField(label='Motivation', required=True, max_length=500,
+      widget=forms.Textarea(attrs={'cols': 50, 'placeholder': 'Explain why the web needs this change. It may be useful to describe what web developers are forced to do without it. When possible, include links to back up your claims in the explainer.', 'maxlength': 500}))
+
+  explainer_links = forms.CharField(label='Explainer link(s)', required=True,
```
Comment (ebidel) [inline models.py]: Should this be required if, for some cases, only a blink-dev email will have gone out?

File: models.py [hunk 5]
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
...
+      help_text='One URL per line.')
+
+  intent_to_implement_url = forms.URLField(required=False, label='Intent to Implement link',
+      help_text='Link to the "Intent to Implement" discussion thread.')
+
+  origin_trial_feedback_url = forms.URLField(required=False, label='Origin Trial feedback summary',
+      help_text='If your feature was available as an Origin Trial, link to a summary of usage and developer feedback. If not, leave this empty.')
+
+  doc_links = forms.CharField(label='Doc link(s)', required=False, max_length=500,
+      widget=forms.Textarea(attrs={'rows': 4, 'cols': 50, 'placeholder': 'Links to design doc(s), if and when available. [This is not required to send out an intent to implement. Please update the intent thread with the design doc when ready]. An explainer and/or design doc is sufficient to start this process.\n\n[Note: Please include links and data, where possible, to support any claims.]', 'maxlength': 500}),
```
Comment (ebidel) [inline models.py]: intent to implement -> Intent to Implement

File: models.py [hunk 6]
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
...
+
+  tag_review = forms.CharField(label='TAG Review', required=True,
+      widget=forms.Textarea(attrs={'rows': 2, 'cols': 50, 'placeholder': 'Link(s) to TAG review(s), or explanation why this is not needed.', 'maxlength': 500}),
+      help_text='One URL per line, optionally followed by a short title, or rationale for omitting TAG review.')
+
+  interop_compat_risks = forms.CharField(label='Interoperability and Compatibility Risks', required=True, max_length=500,
+      widget=forms.Textarea(attrs={'cols': 50, 'maxlength': 500}),
+      help_text='Describe the degree of <a target="_blank" href="https://sites.google.com/a/chromium.org/dev/blink?pli=1#TOC-Policy-for-shipping-and-removing-web-platform-API-features">interoperability risk</a>. For a new feature, the main risk is that it fails to become an interoperable part of the web platform if other browsers do not implement it. For a removal, please review our <a target="_blank" href="https://docs.google.com/document/d/1RC-pBBvsazYfCNNUSkPqAVpSpNJ96U8trhNkfV0v9fk/edit">principles of web compatibility</a>.')
```
Comment (ebidel) [inline models.py]: Not sure this will work. Do the links render properly in the textarea?

File: static/js/admin/feature_form.js [hunk 7]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+const INTENT_IMPLEMENT = 1;
+const INTENT_EXPERIMENT = 2;
+const INTENT_EXTEND_TRIAL = 3;
+const INTENT_IMPLEMENT_SHIP = 4;
+const INTENT_SHIP = 5;
+const INTENT_REMOVE = 6;
+
+let INTENT_IDENTIFIER_NAMES = {};
+
+INTENT_IDENTIFIER_NAMES[INTENT_IMPLEMENT] =      "INTENT_IMPLEMENT";
```
Comment (ebidel) [inline static/js/admin/feature_form.js]: remove the extra alignment spaces throughout.

File: static/js/admin/feature_form.js [hunk 8]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+INTENT_IDENTIFIER_NAMES[INTENT_EXPERIMENT] =     "INTENT_EXPERIMENT";
+INTENT_IDENTIFIER_NAMES[INTENT_EXTEND_TRIAL] =   "INTENT_EXTEND_TRIAL";
+INTENT_IDENTIFIER_NAMES[INTENT_IMPLEMENT_SHIP] = "INTENT_IMPLEMENT_SHIP";
```
Comment (ebidel) [inline static/js/admin/feature_form.js]: single quotes

File: static/js/admin/feature_form.js [hunk 9]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+INTENT_IDENTIFIER_NAMES[INTENT_SHIP] =           "INTENT_SHIP";
+INTENT_IDENTIFIER_NAMES[INTENT_REMOVE] =         "INTENT_REMOVE";
+
+// A object graph mapping form fields to implementation status and whehter or
```
Comment (ebidel) [inline static/js/admin/feature_form.js]:
```suggestion
// An object graph mapping form fields to implementation status and whether or
```

File: static/js/admin/feature_form.js [hunk 10]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+  },
+  'origin_trial_feedback_url': {
+    INTENT_IMPLEMENT:      HIDDEN,
+    INTENT_EXPERIMENT:     HIDDEN,
+    INTENT_EXTEND_TRIAL:   HIDDEN,
+    INTENT_IMPLEMENT_SHIP: HIDDEN,
+    INTENT_SHIP:           VISIBLE_OPTIONAL,
+    INTENT_REMOVE:         VISIBLE_OPTIONAL
+  },
+}
```
Comment (ebidel) [inline static/js/admin/feature_form.js]:
```suggestion
};
```

File: static/js/admin/feature_form.js [hunk 11]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+
+function intentStageChanged(stage) {
+  function set_required(id, required) {
```
Comment (ebidel) [inline static/js/admin/feature_form.js]:
```suggestion
  function setRequired(id, required) {
```

File: static/js/admin/feature_form.js [hunk 12]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+  }
+
+  function show(id) {
+    // Clear the inline style for the control's table row.
+    document.querySelector('#id_' + id).parentNode.parentNode.style = "";
+  }
+
+  function hide(id) {
+    // Set the control's table row style to display: none.
+    document.querySelector('#id_' + id).parentNode.parentNode
```
Comment (ebidel) [inline static/js/admin/feature_form.js]:
```suggestion
    document.querySelector('#id_' + id).parentNode.parentNode.style.display = 'none';
```

File: static/js/admin/feature_form.js [hunk 13]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+	.style = "display: none;";
+  }
+
+  let stage_index = INTENT_IDENTIFIER_NAMES[Number(stage.value)];
```
Comment (ebidel) [inline static/js/admin/feature_form.js]: vars should be camelcase

```suggestion
  let stageIndex = INTENT_IDENTIFIER_NAMES[Number(stage.value)];
```

File: static/js/admin/feature_form.js [hunk 14]
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
...
+    }
+
+    if (FORM_FIELD_GRAPH[id][stage_index][FORM_FIELD_REQUIRED_POS]) {
+      set_required(id, true);
+    } else {
+      set_required(id, false);
+    }
+  }
+
+  let intent_stage_name = document.querySelector('#id_intent_stage_name')
```
Comment (ebidel) [inline static/js/admin/feature_form.js]:
```suggestion
  let intentStageName = document.querySelector('#id_intent_stage_name');
```

File: models.py (hunk 6 above)
Comment (jstenback) [reply models.py]: Yep, the help text ends up below the actual textarea (as opposed to the "placeholder" text which can not contain markup AFAIK).

**Comment by jstenback** [comment]
I believe the update to this pull request addresses all the feedback thus far, and also takes a stab at fixing the issues the linter pointed out as well. Additional feedback always welcome!

File: templates/feature.html [hunk 15]
```diff
@@ -231,6 +231,14 @@ <h3>{% if feature.owner|length == 1 %}Owner{% else %}Owners{% endif %}</h3>
   </section>
   {% endif %}
 
+  {% if feature.intent_to_implement_url %}
+  <section id="intent_to_implement_url">
```
[the remark this answers is not in the document]
Comment (jstenback) [reply templates/feature.html]: It is the presence of a long help_text content that makes the table go so wide. I.e. if there's more help_text than the textbox is wide, then the text box width follows the width of the help_text content. Ironically I just went through and moved all the placeholder content out to the help_text to reduce the redundancy, which I think we should do. The alternative would be to move more things into placeholders, which has the unfortunate effect of making the help text invisible when you go back and want to edit a field that already has some content.

Wide textareas seem like a reasonable trade off between the two IMO, but open to other ideas.

File: static/js/admin/feature_form.js [hunk 16]
```diff
@@ -19,8 +19,8 @@ form.addEventListener('change', function(e) {
     case 'select':
       if (e.target.id === 'id_impl_status_chrome') {
         toggleMilestones(e.target);
-      } else if (e.target.id === 'id_standardization') {
-        toggleSpecLink(e.target);
+      } else if (e.target.id === 'id_intent_stage') {
```
[the remark this answers is not in the document]
Comment (jstenback) [reply static/js/admin/feature_form.js]: Yep, we ask people to send Intent to Implement messages before we ask them to send Intent to Experiment, so the Intent to Experiment needs a link to the Intent to Implement.

File: models.py [hunk 17]
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
...
+
+  debuggability = forms.CharField(label='Debuggability', required=False,
+      widget=forms.Textarea(attrs={'cols': 50, 'placeholder': 'Please describe the desired DevTools debugging support for your feature. Consider emailing https://groups.google.com/forum/?fromgroups#!forum/google-chrome-developer-tools to make sure developers are able to debug and inspect code that uses your feature. For new language features in V8 specifically, refer to the debugger support checklist. If your feature doesn\'t require changes to DevTools in order to provide a good debugging experience, feel free to delete this section.', 'maxlength': 500}),
+      help_text='Description of the desired DevTools debugging support for your feature. Consider emailing the <a href="https://groups.google.com/forum/?fromgroups#!forum/google-chrome-developer-tools">google-chrome-developer-tools</a> list for additional help.')
+
+  all_platforms = forms.BooleanField(required=False, initial=False, label='Supported on all platforms?', help_text='Will this feature be supported on all six Blink platforms (Windows, Mac, Linux, Chrome OS, Android, and Android WebView)?')
```
[the remark this answers is not in the document]
Comment (jstenback) [reply models.py]: Yep, good point. I ended up deleting all the placeholder text and consistently (with 4 small exceptions) using help_text instead. That way the text is always visible, even when a user goes back to edit a feature (placeholder text of course disappears as soon as there's any content in a textarea). This is now consistent with other form field types as well, but has the downside of making the form fields wider.

File: models.py [hunk 18]
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
...
+
+  all_platforms_descr = forms.CharField(label='Platfor Support Explanation', required=False,
```
[the remark this answers is not in the document]
Comment (jstenback) [reply models.py]: Fixed.

File: models.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ebidel) [reply models.py]: The table is set to `width: 100%` right now. You could try this in forms.scss:

```
input[type="text"], input[type="url"], input[type="email"], textarea {
    width: 100%;
    max-width: $max-content-width;
  }
```

and see if that works a little better. Huge inputs look pretty daunting IMO.

[the remark this answers is not in the document]
Comment (jstenback) [reply models.py]: Tried a few things before I got that to do what I wanted, pushed an update to reflect something that seems reasonable here.

**Comment by ebidel** [comment]
Think this looks good. I'm a little worried about adding more required fields, but I guess the intension of this change is to firm up the process.

**Comment by ebidel** [comment]
LMK when you're ready to merge.

**Comment by jstenback** [comment]
Merged per in-person discussion with ebidel.

**Pull request merged**